src/main/java/io/zksync/wallet/DefaultZkSyncWallet.java

```diff
@@ -208,66 +206,98 @@ public Order buildSignedOrder(String recipient, Token sell, Token buy, Tuple2<Bi
         return zkSigner.signOrder(order);
     }
 
+    @Override
+    @SneakyThrows
+    public Order buildSignedLimitOrder(String recipient, Token sell, Token buy, Tuple2<BigInteger, BigInteger> ratio,
+            Integer nonce, TimeRange timeRange) {
+        return this.buildSignedOrder(recipient, sell, buy, ratio, BigInteger.ZERO, nonce, timeRange);
+    }
+
     @Override
     public AccountState getState() {
         return provider.getState(ethSigner.getAddress());
     }
 
     @Override
     public boolean isSigningKeySet() {
-        return this.pubKeyHash.equals(this.zkSigner.getPublicKeyHash());
+        return Objects.equals(this.getPubKeyHash(), this.zkSigner.getPublicKeyHash());
+    }
+
+    @Override
+    public Integer getAccountId() {
+        if (this.accountId == null) {
+            this.loadAccountInfo();
+        }
+
+        return this.accountId;
+    }
+
+    @Override
+    public String getPubKeyHash() {
+        if (this.pubKeyHash == null) {
+            this.loadAccountInfo();
+        }
+
+        return this.pubKeyHash;
+    }
+
+    @Override
+    public EthereumProvider createEthereumProvider(Web3j web3j, ContractGasProvider contractGasProvider) {
+        String contractAddress = this.provider.contractAddress().getMainContract();
+        ZkSync contract = ZkSync.load(contractAddress, web3j, this.ethSigner.getTransactionManager(), contractGasProvider);
+        DefaultEthereumProvider ethereum = new DefaultEthereumProvider(web3j, this.ethSigner, contract);
+        return ethereum;
+    }
+
+    @Override
+    public String getAddress() {
+        return this.ethSigner.getAddress();
     }
 
     @SneakyThrows
-    private SignedTransaction<ChangePubKey<ChangePubKeyECDSA>> buildSignedChangePubKeyTxSigned(TransactionFee fee, Integer nonce,
-            TimeRange timeRange) {
+    private SignedTransaction<ChangePubKey<A>> buildSignedChangePubKeyTx(TransactionFee fee, Integer nonce,
```

Comment:
The clients should be able to do the signing of a transaction and sending it separately